File: backend/app/schemas/audit.py
"""Audit trail schemas for API responses and requests."""

from datetime import datetime, date
from typing import Optional, List, Any, Dict

from pydantic import BaseModel, Field, field_validator

from app.models.enums import AuditAction


# ============================================================================
# Audit Log Schemas
# ============================================================================


class AuditLogResponse(BaseModel):
    """Response schema for an audit log entry."""

    id: int
    table_name: str
    record_id: int
    action: AuditAction
    old_values: Optional[Dict[str, Any]] = None
    new_values: Optional[Dict[str, Any]] = None
    user_id: Optional[int] = None
    username: Optional[str] = None
    timestamp: datetime
    ip_address: Optional[str] = None
    reason: Optional[str] = None
    annotation_count: int = 0

    model_config = {"from_attributes": True}


class AuditLogListResponse(BaseModel):
    """Paginated list of audit log entries."""

    items: List[AuditLogResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class AuditLogFilters(BaseModel):
    """Filters for audit log queries."""

    table_name: Optional[str] = None
    record_id: Optional[int] = None
    action: Optional[AuditAction] = None
    user_id: Optional[int] = None
    date_from: Optional[date] = None
    date_to: Optional[date] = None


# ============================================================================
# Audit Annotation Schemas
# ============================================================================


class AuditAnnotationCreate(BaseModel):
    """Schema for creating an audit annotation."""

    comment: Optional[str] = Field(None, max_length=10000)

    @field_validator("comment")
    @classmethod
    def validate_comment(cls, v):
        if v is not None and v.strip() == "":
            return None
        return v


class AuditAnnotationResponse(BaseModel):
    """Response schema for an audit annotation."""

    id: int
    audit_log_id: int
    user_id: int
    username: Optional[str] = None
    comment: Optional[str] = None
    attachment_filename: Optional[str] = None
    attachment_size: Optional[int] = None
    attachment_hash: Optional[str] = None
    created_at: datetime

    model_config = {"from_attributes": True}


class AuditAnnotationListResponse(BaseModel):
    """List of audit annotations."""

    items: List[AuditAnnotationResponse]
    total: int


# ============================================================================
# Audit Export Schemas
# ============================================================================


class AuditExportRequest(BaseModel):
    """Request schema for exporting audit logs."""

    table_name: Optional[str] = None
    record_id: Optional[int] = None
    action: Optional[AuditAction] = None
    user_id: Optional[int] = None
    date_from: Optional[date] = None
    date_to: Optional[date] = None
    format: str = Field("csv", pattern="^(csv|pdf)$")


# ============================================================================
# Audit Trail Display Schemas (for frontend)
# ============================================================================


class FieldChange(BaseModel):
    """Schema for a single field change."""

    field: str
    old_value: Optional[Any] = None
    new_value: Optional[Any] = None
    display_old: Optional[str] = None
    display_new: Optional[str] = None


class AuditEntryDisplay(BaseModel):
    """Display-ready audit entry for frontend."""

    id: int
    action: str
    action_display: str
    timestamp: datetime
    timestamp_display: str
    username: str
    changes: List[FieldChange]
    reason: Optional[str] = None
    is_bulk_operation: bool = False
    bulk_summary: Optional[str] = None
    annotation_count: int = 0


class AuditTrailResponse(BaseModel):
    """Full audit trail for a record."""

    table_name: str
    record_id: int
    entries: List[AuditEntryDisplay]
    total: int


# ============================================================================
# Archive Schemas (kept for backwards compatibility)
# ============================================================================


class ArchivedLotResponse(BaseModel):
    """Archived (completed) lot response for list view."""

    lot_id: int
    product_id: int
    reference_number: str
    lot_number: str
    product_name: str
    brand: str
    flavor: Optional[str] = None
    size: Optional[str] = None
    status: str  # "released" or "rejected"
    completed_at: datetime  # Released or rejected date
    customer_name: Optional[str] = None
    rejection_reason: Optional[str] = None

    model_config = {"from_attributes": True}


class ArchivedLotsListResponse(BaseModel):
    """Paginated list of archived lots."""

    items: List[ArchivedLotResponse]
    total: int
    page: int
    page_size: int
    total_pages: int

File: frontend/src/api/audit.ts
/**
 * Audit trail API client
 */

import { api } from "./client";

// Types
export type AuditAction =
  | "insert"
  | "update"
  | "delete"
  | "approve"
  | "reject"
  | "validation_failed";

export interface FieldChange {
  field: string;
  old_value: unknown;
  new_value: unknown;
  display_old: string | null;
  display_new: string | null;
}

export interface AuditEntryDisplay {
  id: number;
  action: AuditAction;
  action_display: string;
  timestamp: string;
  timestamp_display: string;
  username: string;
  changes: FieldChange[];
  reason: string | null;
  is_bulk_operation: boolean;
  bulk_summary: string | null;
  annotation_count: number;
}

export interface AuditTrailResponse {
  table_name: string;
  record_id: number;
  entries: AuditEntryDisplay[];
  total: number;
}

export interface AuditAnnotation {
  id: number;
  audit_log_id: number;
  user_id: number;
  username: string | null;
  comment: string | null;
  attachment_filename: string | null;
  attachment_size: number | null;
  attachment_hash: string | null;
  created_at: string;
}

export interface AuditAnnotationListResponse {
  items: AuditAnnotation[];
  total: number;
}

export interface AuditLogFilters {
  table_name?: string;
  record_id?: number;
  action?: AuditAction;
  user_id?: number;
  date_from?: string;
  date_to?: string;
}

/** Metadata for export filename generation */
export interface ExportMetadata {
  brand?: string;
  productName?: string;
  lotNumber?: string;
}

/** View mode for audit trail display */
export type AuditViewMode = "summary" | "detailed";

/** Consolidated row for summary view */
export interface ConsolidatedRow {
  id: string; // unique key for React
  auditIds: number[]; // all audit IDs included in this consolidated row
  primaryAuditId: number; // main audit ID for annotations
  timestamp: Date;
  timestampDisplay: string;
  username: string;
  action: string;
  actionDisplay: string;
  field: string; // field name or description (e.g., "E. coli (6 fields)", "Status")
  oldValue: string; // old value or "—" for creates
  newValue: string; // new value
  reason: string | null;
  annotationCount: number; // total across all consolidated entries
  consolidatedCount: number; // number of entries consolidated (1 if not consolidated)
}

// API functions

/**
 * Get audit trail for a specific record
 */
export async function getAuditTrail(
  tableName: string,
  recordId: number
): Promise<AuditTrailResponse> {
  const response = await api.get<AuditTrailResponse>(
    `/audit/${tableName}/${recordId}/trail`
  );
  return response.data;
}

/**
 * Get annotations for an audit log entry
 */
export async function getAnnotations(
  auditId: number
): Promise<AuditAnnotationListResponse> {
  const response = await api.get<AuditAnnotationListResponse>(
    `/audit/${auditId}/annotations`
  );
  return response.data;
}

/**
 * Add a comment to an audit log entry
 */
export async function addAnnotationComment(
  auditId: number,
  comment: string
): Promise<AuditAnnotation> {
  const params = new URLSearchParams();
  params.append("comment", comment);

  const response = await api.post<AuditAnnotation>(
    `/audit/${auditId}/annotations?${params.toString()}`
  );
  return response.data;
}

/**
 * Add an attachment to an audit log entry
 */
export async function addAnnotationAttachment(
  auditId: number,
  file: File,
  comment?: string
): Promise<AuditAnnotation> {
  const formData = new FormData();
  formData.append("file", file);
  if (comment) {
    formData.append("comment", comment);
  }

  const response = await api.post<AuditAnnotation>(
    `/audit/${auditId}/annotations`,
    formData,
    {
      headers: {
        "Content-Type": "multipart/form-data",
      },
    }
  );
  return response.data;
}

/**
 * Download an attachment
 */
export async function downloadAttachment(
  auditId: number,
  annotationId: number,
  filename: string
): Promise<void> {
  const response = await api.get(
    `/audit/${auditId}/annotations/${annotationId}/download`,
    {
      responseType: "blob",
    }
  );

  // Create download link
  const url = window.URL.createObjectURL(new Blob([response.data]));
  const link = document.createElement("a");
  link.href = url;
  link.setAttribute("download", filename);
  document.body.appendChild(link);
  link.click();
  link.remove();
  window.URL.revokeObjectURL(url);
}

/**
 * Sanitize string for filename (remove spaces and special characters)
 */
function sanitizeForFilename(s: string): string {
  return s.replace(/\s+/g, "").replace(/[^a-zA-Z0-9-_]/g, "");
}

/**
 * Export audit logs as CSV
 */
export async function exportAuditCsv(
  filters?: AuditLogFilters,
  metadata?: ExportMetadata
): Promise<void> {
  const params = new URLSearchParams();
  if (filters?.table_name) params.append("table_name", filters.table_name);
  if (filters?.record_id) params.append("record_id", String(filters.record_id));
  if (filters?.action) params.append("action", filters.action);
  if (filters?.user_id) params.append("user_id", String(filters.user_id));
  if (filters?.date_from) params.append("date_from", filters.date_from);
  if (filters?.date_to) params.append("date_to", filters.date_to);

  const response = await api.get(`/audit/export/csv?${params.toString()}`, {
    responseType: "blob",
  });

  // Generate filename from metadata if available
  const filename =
    metadata?.brand && metadata?.productName && metadata?.lotNumber
      ? `${sanitizeForFilename(metadata.brand)}-${sanitizeForFilename(metadata.productName)}-${sanitizeForFilename(metadata.lotNumber)}.csv`
      : `audit_export_${new Date().toISOString().slice(0, 10)}.csv`;

  // Create download link
  const url = window.URL.createObjectURL(new Blob([response.data]));
  const link = document.createElement("a");
  link.href = url;
  link.setAttribute("download", filename);
  document.body.appendChild(link);
  link.click();
  link.remove();
  window.URL.revokeObjectURL(url);
}

/**
 * Export audit logs as PDF
 */
export async function exportAuditPdf(
  filters?: AuditLogFilters,
  metadata?: ExportMetadata
): Promise<void> {
  const params = new URLSearchParams();
  if (filters?.table_name) params.append("table_name", filters.table_name);
  if (filters?.record_id) params.append("record_id", String(filters.record_id));
  if (filters?.action) params.append("action", filters.action);
  if (filters?.user_id) params.append("user_id", String(filters.user_id));
  if (filters?.date_from) params.append("date_from", filters.date_from);
  if (filters?.date_to) params.append("date_to", filters.date_to);

  const response = await api.get(`/audit/export/pdf?${params.toString()}`, {
    responseType: "blob",
  });

  // Generate filename from metadata if available
  const filename =
    metadata?.brand && metadata?.productName && metadata?.lotNumber
      ? `${sanitizeForFilename(metadata.brand)}-${sanitizeForFilename(metadata.productName)}-${sanitizeForFilename(metadata.lotNumber)}.pdf`
      : `audit_export_${new Date().toISOString().slice(0, 10)}.pdf`;

  // Create download link
  const url = window.URL.createObjectURL(new Blob([response.data]));
  const link = document.createElement("a");
  link.href = url;
  link.setAttribute("download", filename);
  document.body.appendChild(link);
  link.click();
  link.remove();
  window.URL.revokeObjectURL(url);
}
